src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1354,13 +1361,31 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
             try
             {
                 // Send headers
-                http2Stream = await SendHeadersAsync(request, cancellationToken).ConfigureAwait(false);
-
-                // Send request body, if any
-                await http2Stream.SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                bool shouldExpectContinue = request.Content != null && request.HasHeaders && request.Headers.ExpectContinue == true;
 
-                // Wait for response headers to be read.
-                await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                http2Stream = await SendHeadersAsync(request, cancellationToken, flush: shouldExpectContinue).ConfigureAwait(false);
+                if (shouldExpectContinue)
+                {
+                    // Send header and wait a little bit to see if server sends 100, reject code or nothing.
+                    if (NetEventSource.IsEnabled) Trace($"Request content is not null, start processing 100-Continue.");
+                    await http2Stream.SendRequestBodyWithExpect100ContinueAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // Send request body, if any
+                    Task bodyTask = http2Stream.SendRequestBodyAsync(cancellationToken);
+                    if (!request.SendConcurent)
+                    {
+                        await bodyTask.ConfigureAwait(false);
+                        bodyTask = null;
+                    }
+                    // Wait for response headers to be read.
+                    await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                    if (bodyTask != null && bodyTask.IsCompleted && bodyTask.IsFaulted)
```

Comment:
You don't need both the IsCompleted and IsFaulted checks here.  In fact, you probably explicitly don't want the IsFaulted check, in case the task ended as Canceled.  Better to just do `if (bodyTask != null && bodyTask.IsCompleted`.